barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_trace.cpp

```diff
@@ -1568,38 +1568,135 @@ void AvmTraceBuilder::op_nullifier_exists(uint32_t note_offset, uint32_t dest_of
+        .avm_main_ib = read_slot.val,
+        .avm_main_ind_a = dest_offset_is_indirect ? dest_offset : 0,
+        .avm_main_ind_op_a = FF(static_cast<uint32_t>(dest_offset_is_indirect)),
+        .avm_main_internal_return_ptr = FF(internal_return_ptr),
+        .avm_main_mem_idx_a = FF(direct_dest_offset),
+        .avm_main_mem_idx_b = FF(slot_offset),
+        .avm_main_mem_op_a = FF(1),
+        .avm_main_mem_op_b = FF(1),
+        .avm_main_pc = pc, // No PC increment here since this is the same opcode as the rows created below
+        .avm_main_r_in_tag = FF(static_cast<uint32_t>(AvmMemoryTag::FF)),
+        .avm_main_w_in_tag = FF(static_cast<uint32_t>(AvmMemoryTag::FF)),
+    });
+    clk++;
 
-    main_trace.push_back(row);
-    side_effect_counter++;
+    for (uint32_t i = 0; i < size; i++) {
+        FF value = execution_hints.get_side_effect_hints().at(side_effect_counter);
+
+        mem_trace_builder.write_into_memory(
+            call_ptr, clk, IntermRegister::IA, direct_dest_offset + i, value, AvmMemoryTag::FF, AvmMemoryTag::FF);
+
+        auto row = Row{
+            .avm_main_clk = clk,
+            .avm_main_ia = value,
+            .avm_main_ib = read_slot.val + i, // slot increments each time
+            .avm_main_internal_return_ptr = internal_return_ptr,
+            .avm_main_mem_idx_a = direct_dest_offset + i,
+            .avm_main_mem_op_a = 1,
+            .avm_main_pc = pc, // No PC increment here since this is the same opcode for all loop iterations
+            .avm_main_q_kernel_output_lookup = 1,
+            .avm_main_r_in_tag = static_cast<uint32_t>(AvmMemoryTag::FF),
+            .avm_main_rwa = 1,
+            .avm_main_sel_op_sload = FF(1),
+            .avm_main_w_in_tag = static_cast<uint32_t>(AvmMemoryTag::FF),
+        };
+
+        // Output storage read to kernel outputs (performs lookup)
+        kernel_trace_builder.op_sload(clk, side_effect_counter, row.avm_main_ib, row.avm_main_ia);
+
+        // Constrain gas cost
+        gas_trace_builder.constrain_gas_lookup(clk, OpCode::SLOAD);
+
+        main_trace.push_back(row);
+        side_effect_counter++;
+        clk++;
+    }
+    pc++;
 }
 
-void AvmTraceBuilder::op_sstore(uint32_t slot_offset, uint32_t value_offset)
+void AvmTraceBuilder::op_sstore(uint8_t indirect, uint32_t src_offset, uint32_t size, uint32_t slot_offset)
 {
-    auto const clk = static_cast<uint32_t>(main_trace.size()) + 1;
+    auto clk = static_cast<uint32_t>(main_trace.size()) + 1;
+    bool src_offset_is_indirect = is_operand_indirect(indirect, 0);
 
-    Row row =
-        create_kernel_output_opcode_with_metadata(clk, value_offset, AvmMemoryTag::FF, slot_offset, AvmMemoryTag::FF);
-    kernel_trace_builder.op_sstore(clk, side_effect_counter, row.avm_main_ib, row.avm_main_ia);
-    row.avm_main_sel_op_sstore = FF(1);
+    // Resolve loads and indirect
+    auto direct_src_offset = src_offset;
+    if (src_offset_is_indirect) {
```

Comment:
can dest not also be indirect?

Reply:
ok noir will never generate this

Reply:
Yes, but noir won't generate that. Can do in a followup